receiver/vcenterreceiver/client.go

```diff
@@ -0,0 +1,235 @@
+	rps, err := vc.finder.VirtualMachineList(ctx, "*")
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve resource pools: %w", err)
+	}
+	return rps, err
+}
+
+// VSANCluster returns back vSAN performance metrics for a cluster reference
+func (vc *vcenterClient) VSANCluster(
+	ctx context.Context,
+	clusterRef *vt.ManagedObjectReference,
+	startTime time.Time,
+	endTime time.Time,
+) (*[]types.VsanPerfEntityMetricCSV, error) {
+	// not all vCenters support vSAN so just return an empty result
+	if vc.vsanDriver == nil {
+		return &[]types.VsanPerfEntityMetricCSV{}, nil
+	}
+
+	querySpec := []types.VsanPerfQuerySpec{
+		{
+			EntityRefId: "cluster-domclient:*",
+			StartTime:   &startTime,
+			EndTime:     &endTime,
+		},
+	}
+	return vc.queryVsan(ctx, clusterRef, querySpec)
+}
+
+// VSANHosts returns back vSAN performance metrics for a host reference
+// will return nil
+func (vc *vcenterClient) VSANHosts(
+	ctx context.Context,
+	clusterRef *vt.ManagedObjectReference,
+	startTime time.Time,
+	endTime time.Time,
+) (*[]types.VsanPerfEntityMetricCSV, error) {
+	// not all vCenters support vSAN so just return an empty result
+	if vc.vsanDriver == nil {
+		return &[]types.VsanPerfEntityMetricCSV{}, nil
+	}
+	querySpec := []types.VsanPerfQuerySpec{
+		{
+			EntityRefId: "host-domclient:*",
+			StartTime:   &startTime,
+			EndTime:     &endTime,
+		},
+	}
+	return vc.queryVsan(ctx, clusterRef, querySpec)
+}
+
+// VSANHosts returns back vSAN performance metrics for a virtual machine reference
+func (vc *vcenterClient) VSANVirtualMachines(
+	ctx context.Context,
+	clusterRef *vt.ManagedObjectReference,
+	startTime time.Time,
+	endTime time.Time,
+) (*[]types.VsanPerfEntityMetricCSV, error) {
+	// not all vCenters support vSAN so just return an empty result
+	if vc.vsanDriver == nil {
+		return &[]types.VsanPerfEntityMetricCSV{}, nil
+	}
+
+	querySpec := []types.VsanPerfQuerySpec{
+		{
+			EntityRefId: "virtual-machine:*",
+			StartTime:   &startTime,
+			EndTime:     &endTime,
+		},
+	}
+	return vc.queryVsan(ctx, clusterRef, querySpec)
+}
+
+type perfSampleResult struct {
+	counters map[string]*vt.PerfCounterInfo
+	results  []performance.EntityMetric
+}
+
+func (vc *vcenterClient) performanceQuery(
+	ctx context.Context,
+	spec vt.PerfQuerySpec,
+	names []string,
+	objs []vt.ManagedObjectReference,
+) (*perfSampleResult, error) {
+	mgr := performance.NewManager(vc.vimDriver)
+	mgr.Sort = true
+	sample, err := mgr.SampleByName(ctx, spec, names, objs)
+	if err != nil {
+		return nil, err
+	}
+	result, err := mgr.ToMetricSeries(ctx, sample)
+	if err != nil {
+		return nil, err
+	}
+	counterInfoByName, err := mgr.CounterInfoByName(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &perfSampleResult{
+		counters: counterInfoByName,
+		results:  result,
+	}, nil
+}
+
+func (vc *vcenterClient) queryVsan(
+	ctx context.Context,
+	ref *vt.ManagedObjectReference,
+	qs []types.VsanPerfQuerySpec,
+) (*[]types.VsanPerfEntityMetricCSV, error) {
```

Comment:
Why return a pointer to a slice?

Aside from this being unusual, it prevents this function from just being `return vc.vsanDriver.VsanPerfQueryPerf(ctx, ref, qs)`